subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -187,67 +187,280 @@ class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpe
+            }
+        """
+
+        file("src/main/java/Foo.java") << """
+            public class Foo { }
+        """
+
+        def task = ":compileJava"
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        def events = toolchainEvents(task)
+        then:
+        executedAndNotSkipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaCompiler")
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+        events = toolchainEvents(task)
+        then:
+        skipped(task)
+        assertToolchainUsages(events, jdkMetadata, "JavaCompiler")
+    }
+
+    def "emits toolchain usages for compilation that configures java home via fork options pointing outside installations"() {
+        JvmInstallationMetadata jdkMetadata1 = AvailableJavaHomes.getJvmInstallationMetadata(Jvm.current())
+        JvmInstallationMetadata jdkMetadata2 = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+        buildFile << """
+            compileJava {
+                options.forkOptions.javaHome = file("${jdkMetadata2.javaHome}")
+            }
+        """
+
+        file("src/main/java/Foo.java") << """
+            public class Foo { }
+        """
+
+        def task = ":compileJava"
+
+        when:
+        runWithInstallation(jdkMetadata1, task)
+        def events = toolchainEvents(task)
+        then:
+        executedAndNotSkipped(task)
+        assertToolchainUsages(events, jdkMetadata2, "JavaCompiler")
+
+        when:
+        runWithInstallation(jdkMetadata1, task)
+        events = toolchainEvents(task)
+        then:
+        skipped(task)
+        assertToolchainUsages(events, jdkMetadata2, "JavaCompiler")
+    }
+
+    @ToBeImplemented("finding a used toolchain by the path of the launcher executable")
+    @Issue("https://github.com/gradle/gradle/issues/21367")
+    def "emits toolchain usages for test that configures executable path"() {
+        JvmInstallationMetadata jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentVersion)
+
+        buildFile << """
+            compileTestJava {
+                javaCompiler = javaToolchains.compilerFor {
+                    languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+                }
+            }
+
+            def javaExecutable = javaToolchains.launcherFor {
+                languageVersion = JavaLanguageVersion.of(${jdkMetadata.languageVersion.majorVersion})
+            }.get().executablePath
+
+            test {
+                executable = javaExecutable
+            }
+        """
+
+        file("src/test/java/FooTest.java") << """
+            public class FooTest {
+                @org.junit.Test
+                public void test() {}
+            }
+        """
+
+        def task = ":test"
+
+        when:
+        runWithInstallation(jdkMetadata, task)
+//        def events = toolchainEvents(task)
```

Comment:
I think it would be good to assert that there are no events here, so the test fails when we fix this.

Reply:
Done